game-core/src/main/java/games/strategy/triplea/ai/pro/util/MoveBatcher.java

```diff
@@ -0,0 +1,162 @@
+package games.strategy.triplea.ai.pro.util;
+
+import games.strategy.engine.data.Route;
+import games.strategy.engine.data.Unit;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * MoveBatcher implements an algorithm that batches AI moves together, reducing the number of
+ * distinct moves the AI does (and thus the visual delay) by combining moves together.
+ */
+public class MoveBatcher {
+  private static class Move {
+    private final ArrayList<Unit> units;
+    private final Route route;
+    private final HashMap<Unit, Unit> unitsToTransports;
+
+    private Move(
+        final ArrayList<Unit> units,
+        final Route route,
+        final HashMap<Unit, Unit> unitsToTransports) {
+      this.units = units;
+      this.route = route;
+      this.unitsToTransports = unitsToTransports;
+    }
+
+    Move(final ArrayList<Unit> units, final Route route) {
+      this(units, route, null);
+    }
+
+    Move(final Unit unit, final Route route, final Unit transportToLoad) {
+      this(mutableSingletonList(unit), route, new HashMap<>(Map.of(unit, transportToLoad)));
+    }
+
+    private static ArrayList<Unit> mutableSingletonList(final Unit unit) {
+      return new ArrayList<>(Collections.singletonList(unit));
+    }
+
+    boolean isTransportLoad() {
+      return this.unitsToTransports != null;
+    }
+
+    boolean canMergeWith(final Move other) {
+      return other.isTransportLoad() == isTransportLoad() && route.equals(other.route);
+    }
+
+    boolean mergeWith(final Move other) {
+      if (canMergeWith(other)) {
```

Comment:
Is there a clean way to split the "canMerge" concern from the "merge" concern?

It seems like it could make sense for someone to execute this by itself, eg:
```mergetWith(other)``` and ignore the return value. In which case we might have a silent no-op. On the other hand, the usage in an if statement has side effects beyond returning a 'true/false'

Reply:
But this means if someone calls mergeWith() it can result in illegal moves being produced. Are you sure this is what you want? It seems bad for the API to allow this to happen.

Reply:
That sounds like it would be even more brittle.

Making 'mergeWith' functional would be a nice way to go, maybe out of scope/difficult given how much we rely on mutability.

We could make `mergeWith` do a precondition check that `canMerge` is true, from a usage point of view, maybe it's a wash. Reluctantly, we may need for the mutability issues to be fixed before we can fix this one.

You would agree we have a SRP violation? It does seem we have two things wrapped in one.

As a functional variant that does not blow up with precondition, an `Optional<Move>` being returned could be a nice API to indicate we were able to create a valid merge.